consensus/src/detail/tests/reorgs_tests.rs

```diff
@@ -0,0 +1,411 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Author(s): S. Afach, A. Sinitsyn
+
+use std::sync::Mutex;
+
+use crate::detail::tests::test_framework::BlockTestFrameWork;
+use crate::detail::tests::*;
+use blockchain_storage::Store;
+use common::chain::config::create_mainnet;
+
+#[test]
+fn test_reorg_simple() {
+    common::concurrency::model(|| {
+        let config = create_mainnet();
+        let storage = Store::new_empty().unwrap();
+        let mut consensus = Consensus::new_no_genesis(config, storage).unwrap();
+
+        // process the genesis block
+        let result = consensus.process_block(
+            consensus.chain_config.genesis_block().clone(),
+            BlockSource::Local,
+        );
+        assert!(result.is_ok());
+        assert_eq!(
+            consensus
+                .blockchain_storage
+                .get_best_block_id()
+                .expect(ERR_BEST_BLOCK_NOT_FOUND),
+            Some(consensus.chain_config.genesis_block().get_id())
+        );
+
+        // Process the second block
+        let block = produce_test_block(
+            &consensus.chain_config,
+            consensus.chain_config.genesis_block(),
+            false,
+        );
+        let new_id = Some(block.get_id());
+        assert!(consensus.process_block(block, BlockSource::Local).is_ok());
+        assert_eq!(
+            consensus
+                .blockchain_storage
+                .get_best_block_id()
+                .expect(ERR_BEST_BLOCK_NOT_FOUND),
+            new_id
+        );
+
+        // Process the parallel block and choose the better one
+        let block = produce_test_block(
+            &consensus.chain_config,
+            consensus.chain_config.genesis_block(),
+            false,
+        );
+        // let new_id = Some(block.get_id());
+        assert!(consensus.process_block(block.clone(), BlockSource::Local).is_ok());
+        assert_ne!(
+            consensus
+                .blockchain_storage
+                .get_best_block_id()
+                .expect(ERR_BEST_BLOCK_NOT_FOUND),
+            Some(consensus.chain_config.genesis_block().get_id())
+        );
+        assert_eq!(
+            consensus
+                .blockchain_storage
+                .get_best_block_id()
+                .expect(ERR_BEST_BLOCK_NOT_FOUND),
+            new_id
+        );
+
+        // Produce another block that cause reorg
+        let new_block = produce_test_block(&consensus.chain_config, &block, false);
+        let new_id = Some(new_block.get_id());
+        assert!(consensus.process_block(new_block, BlockSource::Local).is_ok());
+        assert_eq!(
+            consensus
+                .blockchain_storage
+                .get_best_block_id()
+                .expect(ERR_BEST_BLOCK_NOT_FOUND),
+            new_id
+        );
+    });
+}
+
+#[test]
+fn test_very_long_reorgs() {
```

Comment:
I don't see a very long reorg test. A very long reorg implies that you create something like 2000 blocks, than you create another chain that's 2001 blocks longer, and then you make sure that it's the new state.

Reply:
OK. Based on the call, we'll do this later.

Reply:
https://github.com/mintlayer/mintlayer-core/issues/173 I made the issue for it